File: 399. Evaluate Division
/Solution.cpp
class Solution {
public:
    unordered_map<string, vector<pair<string, double>>> graph;
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        int q = queries.size(), e = equations.size();
        vector<double> res(q, 0);
        for (int i = 0; i < e; i++) {
            if (graph.find(equations[i][0]) == graph.end()) {
                vector<pair<string, double>> x;
                graph[equations[i][0]] = x;
            }
            if (graph.find(equations[i][1]) == graph.end()) {
                vector<pair<string, double>> y;
                graph[equations[i][1]] = y;
            }
            graph[equations[i][0]].push_back(pair(equations[i][1], values[i]));
            graph[equations[i][1]].push_back(pair(equations[i][0], 1 / values[i]));
        }
        for (int i = 0; i < q; i++) {
            unordered_set<string> visited;
            res[i] = dfs(queries[i][0], queries[i][1], visited, 1);
        }
        return res;
    }

    double dfs(string curr, string target, unordered_set<string> visited, double v) {
        if (visited.find(curr) != visited.end() || graph.find(curr) == graph.end()) return -1;
        if (curr == target) return v;
        visited.insert(curr);
        for (pair<string, double> neighbour : graph[curr]) {
            double foll = dfs(neighbour.first, target, visited, v * neighbour.second);
            if (foll != -1) return foll;
        }
        return -1;
    }
};

File: 399. Evaluate Division
/Solution.py
class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        q = len(queries)
        self.graph, res = dict(), [0 for i in range(q)]
        for i, (x, y) in enumerate(equations):
            if x not in self.graph: self.graph[x] = []
            if y not in self.graph: self.graph[y] = []
            self.graph[x].append((y, values[i]))
            self.graph[y].append((x, 1 / values[i]))
        for i, (x, y) in enumerate(queries):
            res[i] = self.dfs(x, y, set(), 1)
        return res
    def dfs(self, curr, target, visited, v):
        if curr in visited or curr not in self.graph: return -1
        if curr == target: return v
        visited.add(curr)
        for neighbour, w in self.graph[curr]:
            foll = self.dfs(neighbour, target, visited, v * w)
            if foll != -1: return foll
        return -1